src/System.Text.Http.Parser/HttpParser.cs

```diff
@@ -340,6 +340,95 @@ public class HttpParser : IHttpParser
             }
         }
 
+        public unsafe bool ParseHeaders<T>(T handler, ReadOnlyBytes buffer, out int consumedBytes) where T : IHttpHeadersHandler
+        {
+            var span = buffer.First.Span;
+            var rest = buffer.Rest;
+            var length = span.Length;
+            int index = -1;
+            State state = State.ParsingName;
+            int nameStart = 0, nameEnd = 0, valueStart = 0, valueEnd = 0;
+            bool sawCRLF = false;
+            bool straddling = false;
+            fixed (byte* pBuffer = &span.DangerousGetPinnableReference()) {
+                while (true) {
+                    index++;
+                    // get next byte 
+                    while (length <= index) {
+                        if (rest == null) {
+                            RejectRequest(RequestRejectionReason.InvalidRequestHeadersNoCRLF);
+                        }
+                        span = rest.First.Span;
+                        length = span.Length;
+                        rest = rest.Rest;
+                        index = 0;
+                        straddling = true;
+                        throw new NotImplementedException("call to OnHeader below needs to account for straddling spans");
+                    }
+                    var next = pBuffer[index];
+
+                    if (state == State.ParsingValue) {
+                        if (next == ByteCR) {
+                            valueEnd = index;
+                            state = State.ArferCR;
+                        }
+                        // TODO: should value characters be validated?
+                        // TODO: should whitespace after all values (i.e. right before CRLF) be removed here?
+                        // else just keep advancing chracters in name
+                    }
+                    else if (state == State.ParsingName) {
+                        if (next == ByteColon) {
+                            state = State.BeforValue;
+                            nameEnd = index;
```

Comment:
Should this be index - 1? Otherwise, aren't you including the ByteColon as part of the name?